File: solactive-ui/src/blackScholes.js
function normCdf(x) {
  return 0.5 * (1 + erf(x / Math.sqrt(2)));
}

function normPdf(x) {
  return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
}

function erf(x) {
  const sign = x < 0 ? -1 : 1;
  const a1 = 0.254829592;
  const a2 = -0.284496736;
  const a3 = 1.421413741;
  const a4 = -1.453152027;
  const a5 = 1.061405429;
  const p = 0.3275911;
  const t = 1 / (1 + p * Math.abs(x));
  const y = 1 - (((((a5 * t + a4) * t + a3) * t + a2) * t + a1) * t) * Math.exp(-x * x);
  return sign * y;
}

function tauStd(days) {
  return Math.max(days, 0) / 252;
}

function tauCd(days) {
  return Math.max(days, 0) / 365;
}

export function forwardPrice(spot, rate, divYield, daysToExpiry) {
  return spot * Math.exp((rate - divYield) * tauCd(daysToExpiry));
}

export function bsPutPrice(spot, strike, daysToExpiry, rate, divYield, vol) {
  if (daysToExpiry <= 0) return Math.max(strike - spot, 0);
  if (vol <= 0 || spot <= 0 || strike <= 0) return Math.max(strike - spot, 0);

  const tStd = tauStd(daysToExpiry);
  const tCd = tauCd(daysToExpiry);
  const fwd = forwardPrice(spot, rate, divYield, daysToExpiry);
  const d1 = (Math.log(fwd / strike) + 0.5 * vol * vol * tStd) / (vol * Math.sqrt(tStd));
  const d2 = d1 - vol * Math.sqrt(tStd);
  return Math.exp(-rate * tCd) * (strike * normCdf(-d2) - fwd * normCdf(-d1));
}

export function bsVega(spot, strike, daysToExpiry, rate, divYield, vol) {
  if (daysToExpiry <= 0 || vol <= 0) return 0;
  const tStd = tauStd(daysToExpiry);
  const tCd = tauCd(daysToExpiry);
  const fwd = forwardPrice(spot, rate, divYield, daysToExpiry);
  const d1 = (Math.log(fwd / strike) + 0.5 * vol * vol * tStd) / (vol * Math.sqrt(tStd));
  return spot * Math.exp(-divYield * tCd) * normPdf(d1) * Math.sqrt(tStd);
}

export function optionSpread(spot, vol, vega, costFloor, vegaRatioMin, vegaRatioScale, ivBarrier) {
  const ratio = Math.max(vegaRatioMin, (vegaRatioScale * vol) / ivBarrier);
  return spot * Math.max(costFloor, (ratio * vega) / (100 * spot));
}

export function nearestStrike(target, interval) {
  if (interval <= 0) return target;
  const n = Math.round(target / interval);
  const lower = n * interval;
  const upper = (n + 1) * interval;
  return Math.abs(target - lower) <= Math.abs(upper - target) ? lower : upper;
}

export function realizedVol(returns, lookback) {
  if (!returns.length) return null;
  const window = returns.length >= lookback ? returns.slice(-lookback) : returns;
  if (window.length < 2) return null;
  const mean = window.reduce((a, b) => a + b, 0) / window.length;
  const variance = window.reduce((s, r) => s + (r - mean) ** 2, 0) / Math.max(window.length - 1, 1);
  return Math.sqrt(variance) * Math.sqrt(252);
}

export function clampIv(vol, ivMin, ivMax) {
  return Math.min(ivMax, Math.max(ivMin, vol));
}

/** Bisection implied vol; bounded per guideline (0.5% – 500%). */
export function solvePutIv(spot, strike, daysToExpiry, rate, divYield, targetPrice, ivMin, ivMax) {
  const intrinsic = Math.max(strike - spot, 0);
  if (targetPrice <= intrinsic + 1e-8) return ivMin;
  let lo = ivMin;
  let hi = ivMax;
  for (let i = 0; i < 64; i++) {
    const mid = (lo + hi) / 2;
    const p = bsPutPrice(spot, strike, daysToExpiry, rate, divYield, mid);
    if (p > targetPrice) hi = mid;
    else lo = mid;
  }
  return clampIv((lo + hi) / 2, ivMin, ivMax);
}

File: defensive_put/black_scholes.py
import math

import numpy as np


def _norm_cdf(x: float) -> float:
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def _norm_pdf(x: float) -> float:
    return math.exp(-0.5 * x * x) / math.sqrt(2.0 * math.pi)


def _tau_std(days: float) -> float:
    return max(days, 0) / 252.0


def _tau_cd(days: float) -> float:
    return max(days, 0) / 365.0


def forward_price(spot: float, rate: float, div_yield: float, days_to_expiry: float) -> float:
    t = _tau_cd(days_to_expiry)
    return spot * math.exp((rate - div_yield) * t)


def bs_put_price(
    spot: float,
    strike: float,
    days_to_expiry: float,
    rate: float,
    div_yield: float,
    vol: float,
) -> float:
    if days_to_expiry <= 0:
        return max(strike - spot, 0.0)
    if vol <= 0 or spot <= 0 or strike <= 0:
        return max(strike - spot, 0.0)

    t_std = _tau_std(days_to_expiry)
    t_cd = _tau_cd(days_to_expiry)
    fwd = forward_price(spot, rate, div_yield, days_to_expiry)
    sigma = vol
    d1 = (math.log(fwd / strike) + 0.5 * sigma**2 * t_std) / (sigma * math.sqrt(t_std))
    d2 = d1 - sigma * math.sqrt(t_std)
    return math.exp(-rate * t_cd) * (strike * _norm_cdf(-d2) - fwd * _norm_cdf(-d1))


def bs_vega(
    spot: float,
    strike: float,
    days_to_expiry: float,
    rate: float,
    div_yield: float,
    vol: float,
) -> float:
    if days_to_expiry <= 0 or vol <= 0:
        return 0.0
    t_std = _tau_std(days_to_expiry)
    t_cd = _tau_cd(days_to_expiry)
    fwd = forward_price(spot, rate, div_yield, days_to_expiry)
    d1 = (math.log(fwd / strike) + 0.5 * vol**2 * t_std) / (vol * math.sqrt(t_std))
    return spot * math.exp(-div_yield * t_cd) * _norm_pdf(d1) * math.sqrt(t_std)


def option_spread(
    spot: float,
    vol: float,
    vega: float,
    cost_floor: float,
    vega_ratio_min: float,
    vega_ratio_scale: float,
    iv_barrier: float,
) -> float:
    ratio = max(vega_ratio_min, vega_ratio_scale * vol / iv_barrier)
    return spot * max(cost_floor, ratio * vega / 100.0)


def nearest_strike(target: float, interval: float) -> float:
    if interval <= 0:
        return target
    n = round(target / interval)
    lower = n * interval
    upper = (n + 1) * interval
    if abs(target - lower) <= abs(upper - target):
        return lower
    return upper


def realized_vol(returns: np.ndarray, lookback: int) -> float:
    if len(returns) < 2:
        return 0.20
    window = returns[-lookback:] if len(returns) >= lookback else returns
    vol = float(np.std(window, ddof=1)) * math.sqrt(252)
    return float(np.clip(vol, 0.05, 0.80))
